_src/_posts/2017-01-31-learning-rtb.markdown

```diff
@@ -0,0 +1,205 @@
+the expected income too low (thus missing opportunities with $c < \$0.1$) or too
+high (thus buying non-opportunities with $c > \$0.1$)[^risk]. The bottom line is
+that our bidder may be able to pick an optimal pair (bid, creative) for the
+market transaction in course from:
+(i) The estimated conditional conversion rate for the goal event.
+(ii) Our customer valuation of the goal event.
+(iii) The budget constraint for the advertising campaign.
+(iv) A narrow-minded profit maximization behavior.
+
+In order to estimate conversion rates, we implemented a SGD (Stochastic
+Gradient Descent) algorithm or, more specifically, [a FTRL-P][ftrlp] (Follow the
+Regularized Leader Proximal) algorithm with [Adagrad][adagrad] (Adaptive
+Learning Rate) and [ElasticNet][enet] (Lasso and Ridge) regularization. This is
+a state-of-the-art algorithm mostly used in online advertising because of its
+[convenient trade-off][sgd] between optimization and estimation errors in
+scenarios where computational capacity ---and not available data--- is the
+real bottleneck. Every day our system is able to learn from an online stream of
+tens of millions of events published by the bidder into a very
+lightweight middleware implemented on top of [ZMQ][zmq].
+
+The first step after receiving an event from the stream is to turn it into a
+numerical vector that the estimator can consume. Since most of our variables are
+categorical and since we consider many interactions between these variables, the
+output vector will be a sparse binary array of "dummy" variables. We feed this
+vector to a cluster of estimators, each one parameterized according to a grid of
+parameters controlling strength of regularization, number of hashing bits (yes,
+we take advantage of the [hashing trick][trick] as everybody else does),
+estimator memory, etc. Periodically, we evaluate the estimators in the cluster
+using a mix of progressive validation and bootstrap validation in order to
+select the best performing parameterization. The chosen estimator is then
+converted to a tree-like data structure thoroughly optimized for fast model
+evaluation and, finally, it's exported to the bidding subsystem.
+
+We now move on to a more detailed description of the FTRL-P algorithm itself,
+which is the workhorse of our learning system. Let $(x,y)$ denote the input
+vector, with $x$ a set of binary features and $y$ a binary response. Also, let
+$\theta$ denote the vector of model coefficients. Then our prediction at time
+$t$ is computed from the logistic regression model:
+$$
+\hat{y_t} = \frac{1}{1 + e^{-{\theta_t}^T x_t}}
+$$
+Of course, the goal of the algorithm is to learn $\theta_t$ from
+a succession of input events $(x_1,y_1), \ldots, (x_t,y_t)$. For this
+we minimize the sum over each previously seen input event of:
+(i) (The gradient of) the log-likelihood loss, which is a [logloss][] for the
+logistic model.
+(ii) A stabilizing term that introduces some strong convexity into the mix.
```

Comment:
(Only) If items in the list are related, it's better to use ";" instead of "." to enumerate, and leave the "." for the last item (they are part of the same sentence -> they are not capitalized).